### PR TITLE
Ensure stateOK is reported only when all components have sent updates

### LINKED ISSUE #11065: `stateOK` reported in `processState.getState()` before all components have sent updates
### Description

I think that `processState.getState()` can return `stateOK` before it should.

In the following piece of code it seems that `stateOK` can be returned as soon as a single component reports their status as ok, not _all_ of them:
https://github.com/gravitational/teleport/blob/ccbe16181a7c023b803793987a5d54a439aacfcf/lib/service/state.go#L153-L155

Replacing `len(f.states) > 0` with something like `len(f.states) == componentsCount` should fix it.

This was initially noticed in https://github.com/gravitational/cloud/issues/1432.

### PR DESCRIPTION
**Pull request by vitorenesduarte**
Fixes #11065.

This PR:
- ensures  that `TeleportReadyEvent` is only produced when all components that send heartbeats (i.e. call [`process.onHeartbeat`](https://github.com/gravitational/teleport/blob/16bf416556f337b045b66dc9c3f5a3e16f8cc988/lib/service/service.go#L358-L366)) are ready
- changes `TeleportProcess.registerTeleportReadyEvent` so that it returns a count of these components (let's call it `componentCount`)
- uses `componentCount` to also ensure that `stateOK` is only reported when all the components have sent their heartbeat, thus fixing #11065

Since it seems difficult to know when `TeleportProcess.registerTeleportReadyEvent` should be updated, with the goal of quickly detecting a bug when it's introduced we have that:
1. if `componentCount` is lower than it should, then the service fails to start (due to #11725)
2. if `componentCount` is higher than it should, then an error is logged in function `processState.getStateLocked`.

### Testing
None.

### CODE REVIEW

**Comment by vitorenesduarte** [comment]
> @vitorenesduarte I'm not certain, but does `process.ServiceCount()` work?

Thanks! I initially tried this but `process.ServiceCount()` is not exactly what's needed here. There are more services than components (e.g. the auth component has 3 services: "auth.tls", "auth.heartbeat.broadcast", "auth.broadcast", "auth.heartbeat", and "auth.shutdown").

File: lib/service/cfg.go [hunk 1]
```diff
@@ -302,6 +302,45 @@ func (cfg *Config) DebugDumpToYAML() string {
...
+
+	if cfg.SSH.Enabled {
+		componentCount++
+	}
+
+	proxyConfig := cfg.Proxy
+	if proxyConfig.Enabled {
+		componentCount++
+	}
+	if proxyConfig.Kube.Enabled && !proxyConfig.Kube.ListenAddr.IsEmpty() && !proxyConfig.DisableReverseTunnel {
```
Comment (smallinsky) [inline lib/service/cfg.go:321]: How does the `proxyConfig.DisableReverseTunnel` impact `kubernetes_service`  ? 

Also I think that there are some cases where `proxyConfig.Kube.ListenAddr` is empty but kube service should be started

For example when teleport is stared as kube agent: 
```yaml
teleport:
    kubernetes_service:
    enabled: true
    kubeconfig_file: /path/to/kubeconfig

    auth_service:
      enabled: false
    
    ssh_service:
      enabled: false
    
    proxy_service:
      enabled: false
    
    app_service:
      enabled: false
    
    db_service:
      enabled: false
  ```

File: lib/service/state.go [hunk 2]
```diff
@@ -57,9 +57,10 @@ func init() {
 
 // processState tracks the state of the Teleport process.
 type processState struct {
-	process *TeleportProcess
-	mu      sync.Mutex
-	states  map[string]*componentState
+	process        *TeleportProcess
+	mu             sync.Mutex
+	states         map[string]*componentState
+	componentCount int
```
Comment (smallinsky) [inline lib/service/state.go:63]: Can we emphasise in variable name what is the difference between current component calculated based on `len(states`
and desired `stateOK` `componentCount` ?

Comment (vitorenesduarte) [reply lib/service/state.go]: Tried to improve this in 4c0fca2669aa462cbb800dda71d5fbc71e4c381d.

File: lib/service/cfg.go (hunk 1 above)
Comment (vitorenesduarte) [reply lib/service/cfg.go:321]: > How does the `proxyConfig.DisableReverseTunnel` impact `kubernetes_service` ?

I'm not familiar enough with Teleport to able to answer this.

I came up with `if proxyConfig.Kube.Enabled && !proxyConfig.Kube.ListenAddr.IsEmpty() && !proxyConfig.DisableReverseTunnel {` by putting the following two conditions together:

https://github.com/gravitational/teleport/blob/b70c1d6d63ad2c91de27ec527d15a5ceb9df4886/lib/service/service.go#L3043
https://github.com/gravitational/teleport/blob/b70c1d6d63ad2c91de27ec527d15a5ceb9df4886/lib/service/service.go#L2581-L2588

`ComponentCount` counts the number of components that send heartbeats. When the condition above holds, it seems that [the proxy kube component](https://github.com/gravitational/teleport/blob/b70c1d6d63ad2c91de27ec527d15a5ceb9df4886/lib/service/service.go#L3053) will be started and [will send heartbeats](https://github.com/gravitational/teleport/blob/b70c1d6d63ad2c91de27ec527d15a5ceb9df4886/lib/service/service.go#L3080).

> Also I think that there are some cases where `proxyConfig.Kube.ListenAddr` is empty but kube service should be started
> 
> For example when teleport is stared as kube agent:
> 
> ```yaml
> teleport:
>     kubernetes_service:
>     enabled: true
>     kubeconfig_file: /path/to/kubeconfig
> 
>     auth_service:
>       enabled: false
>     
>     ssh_service:
>       enabled: false
>     
>     proxy_service:
>       enabled: false
>     
>     app_service:
>       enabled: false
>     
>     db_service:
>       enabled: false
> ```

I think that in this case this function will correctly return `1` due to this piece of code below:
```go
	if cfg.Kube.Enabled {
		componentCount++
	}
```

**Comment by pierrebeaucamp** [comment]
@smallinsky / @zmb3 / @r0mant can we get another review of this?

**Review by zmb3 — APPROVED** [review]
I don't _love_ it, as there seems to be a lot of guess work involved in figuring out how many components to expect.

Unfortunately, I don't have a better idea at this point.

**Review by smallinsky — APPROVED** [review]

**Comment by vitorenesduarte** [comment]
> @zmb3 @smallinsky Since it seems difficult to have a compile-time error when this new function `Config.ComponentCount` should be updated, maybe we can have the following two run-time errors instead (with the goal of quickly detecting a bug when it's introduced):
> 
> 1. if `Config.ComponentCount` undercounts, then the service fails to start
> 2. if `Config.ComponentCount` overcounts, then an error is logged
> 
> This PR already solves 2. with the new error added to function `processState.getStateLocked`.
> 
> While researching how to solve 1., I noticed #11724. The fix for #11724 in #11725 ensures that a startup error is thrown if the `TeleportReadyEvent` is not emitted.
> 
> So, if #11725 makes sense, we can also solve 1. by making sure that the `Config.ComponentCount` function has _equal (not similar)_ conditions to the ones in the code that creates the `TeleportReadyEvent` event mapping. (We can't ensure this with a compile-time error, but we can leave a note that the two should be updated together.)

With #11725 merged, I pushed 734795f75f1dd9f8d21dfcce0d42035031c21662 so that this PR now implements both 1. and 2. @zmb3 @smallinsky

File: lib/service/service.go [hunk 3]
```diff
@@ -3440,6 +3451,9 @@ func (process *TeleportProcess) waitForAppDepend() {
 
 // registerTeleportReadyEvent ensures that a TeleportReadyEvent is produced
 // when all components have started.
+// Note that this function should be kept in sync with the Config.ComponentCount function so that
```
Comment (espadolini) [inline lib/service/service.go:3454]: If the mapping is 1:1, can't we just calculate the value here and store it in the process?

Comment (vitorenesduarte) [reply lib/service/service.go]: Oh, I think this is a good idea. Implemented something similar in 16b4228daee0a69529d994b7458a363f4a97adb1.

**Review by espadolini — APPROVED** [review]

**Comment by vitorenesduarte** [comment]
Updated the top-level comment to reflect the latest changes.

**Pull request merged**